marimo/_plugins/ui/_impl/table.py

```diff
@@ -346,3 +360,87 @@ def search(self, args: SearchTableArgs) -> Union[JSONType, str]:
         # Save the manager to be used for selection
         self._filtered_manager = result
         return result.limit(TableManager.DEFAULT_ROW_LIMIT).to_data()
+
+    def _apply_formatting(self, data: TableData) -> TableData:
+        # Apply formatting to the given data based on the format mapping.
+        import pandas as pd
+        import polars as pl
+        import pyarrow as pa
+
+        if not self._format_mapping:
+            return data
+
+        if isinstance(data, list | tuple):
+            if all(isinstance(item, dict) for item in data):
+                return [self._format_row(row) for row in data]
+            return [self._format_value("", item) for item in data]
+
+        if isinstance(data, dict):
+            return {
+                col: self._format_column(col, values)
+                for col, values in data.items()
+            }
+
+        if isinstance(data, pd.DataFrame):
+            data_copy = data.copy()
+            for col in data_copy.columns:
+                if col in self._format_mapping:
+                    data_copy[col] = data_copy[col].apply(
+                        lambda x, col=col: self._format_value(col, x)
+                    )
+            return data_copy
+
+        if isinstance(data, pl.DataFrame):
+            data_copy = data.clone()
+            for col in data_copy.columns:
+                if col in self._format_mapping:
+                    formatted_col = pl.Series(
+                        col,
+                        [self._format_value(col, x) for x in data_copy[col]],
+                    )
+                    data_copy = data_copy.with_columns(formatted_col)
+            return data_copy
+
+        if isinstance(data, pa.Table):
+            data_copy = data
+            for col in data_copy.column_names:
+                if col in self._format_mapping:
+                    # Get the column index
+                    col_index = data_copy.schema.get_field_index(col)
+                    # Apply the transformation to each element in the column
+                    transformed_column = pa.array(
+                        [
+                            self._format_value(col, value.as_py())
+                            for value in data_copy[col_index]
+                        ],
+                        type=data_copy.schema.field(col_index).type,
+                    )
+                    # Replace the column in the copy of the table
+                    data_copy = data_copy.set_column(
+                        col_index, col, transformed_column
+                    )
+            return data_copy
+
+        return data
+
+    def _format_row(self, row: Dict[str, JSONType]) -> Dict[str, JSONType]:
```

Comment:
can we bring these utils a file `/marimo/_plugins/ui/_impl/tables/format.py`.

it should be much easier to unit test without having to go through `mo.ui.table`.